internal/controller/apiextensions/usage/reconciler.go

```diff
@@ -0,0 +1,405 @@
+	selectorResolver
+}
+
+// NewReconciler returns a Reconciler of Usages.
+func NewReconciler(mgr manager.Manager, opts ...ReconcilerOption) *Reconciler {
+	kube := unstructured.NewClient(mgr.GetClient())
+
+	r := &Reconciler{
+		client: xpresource.ClientApplicator{
+			Client:     kube,
+			Applicator: xpresource.NewAPIUpdatingApplicator(kube),
+		},
+
+		usage: usageResource{
+			Finalizer:        xpresource.NewAPIFinalizer(kube, finalizer),
+			selectorResolver: newAPISelectorResolver(kube),
+		},
+
+		log:    logging.NewNopLogger(),
+		record: event.NewNopRecorder(),
+
+		pollInterval: 30 * time.Second,
+	}
+
+	for _, f := range opts {
+		f(r)
+	}
+	return r
+}
+
+// A Reconciler reconciles Usages.
+type Reconciler struct {
+	client xpresource.ClientApplicator
+
+	usage usageResource
+
+	log    logging.Logger
+	record event.Recorder
+
+	pollInterval time.Duration
+}
+
+// Reconcile a Usage resource by resolving its selectors, defining ownership
+// relationship, adding a finalizer and handling proper deletion.
+func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) { //nolint:gocyclo // Reconcilers are typically complex.
+	log := r.log.WithValues("request", req)
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	// Get the usageResource resource for this request.
+	u := &v1alpha1.Usage{}
+	if err := r.client.Get(ctx, req.NamespacedName, u); err != nil {
+		log.Debug(errGetUsage, "error", err)
+		return reconcile.Result{}, errors.Wrap(xpresource.IgnoreNotFound(err), errGetUsage)
+	}
+
+	if err := r.usage.resolveSelectors(ctx, u); err != nil {
+		log.Debug(errResolveSelectors, "error", err)
+		err = errors.Wrap(err, errResolveSelectors)
+		r.record.Event(u, event.Warning(reasonResolveSelectors, err))
+		return reconcile.Result{}, err
+	}
+
+	r.record.Event(u, event.Normal(reasonResolveSelectors, "Selectors resolved, if any."))
+
+	of := u.Spec.Of
+	by := u.Spec.By
+
+	// Identify used xp composed as an unstructured object.
+	used := composed.New(composed.FromReference(v1.ObjectReference{
+		Kind:       of.Kind,
+		Name:       of.ResourceRef.Name,
+		APIVersion: of.APIVersion,
+	}))
+
+	if meta.WasDeleted(u) {
+		if by != nil {
+			// Identify using resource as an unstructured object.
+			using := composed.New(composed.FromReference(v1.ObjectReference{
+				Kind:       by.Kind,
+				Name:       by.ResourceRef.Name,
+				APIVersion: by.APIVersion,
+			}))
+			// Get the using resource
+			err := r.client.Get(ctx, client.ObjectKey{Name: by.ResourceRef.Name}, using)
+			if xpresource.IgnoreNotFound(err) != nil {
+				log.Debug(errGetUsing, "error", err)
+				err = errors.Wrap(xpresource.IgnoreNotFound(err), errGetUsing)
+				r.record.Event(u, event.Warning(reasonGetUsing, err))
+				return reconcile.Result{}, err
+			}
+
+			if err == nil {
+				// Using resource is still there, so we need to wait for it to be deleted.
+				msg := "Waiting for using resource to be deleted."
```

Comment:
could we add to the message gvk and name of the resource we are waiting to be deleted?